### PR TITLE
jobs are failed when exceptions raised

### PR DESCRIPTION
**Pull request by jpslav**
Done.  @Dantemss and @joemsak please review.

* When an exception is raised in a routine, fail the job status and add an error to the job status for the exception, then allow the exception to bubble out.
* Fixed an issue where different calls making BackgroundJob status objects were using strings vs symbols (now that I think about it this may only have been an issue locally because dev uses the MemoryStore, whereas other envs use a RedisStore which requires keys to be strings).  In any account, I forced all keys to be strings.
* Fixed an issue where `BackgroundJob.all` was returning duplicate `BackgroundJob` objects.
* Unrelated change to set progress to 1.0 when a job is completed.  (no more 0% and "completed" jobs!)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -100,10 +100,7 @@ end
 
 Additionally, see below for a discussion on how to transfer errors from ActiveRecord models.
 
-Any `StandardError` raised within a routine will be caught and transformed into a fatal error with `:kind` set to `:exception`.  The caller of this routine can choose to reraise this exception by calling `reraise_exception!` on the returned errors object:
-
-    result = MyRoutine.call(42)
-    result.errors.reraise_exception! # does nothing if there were no exception errors
```
Comment (jpslav) [inline README.md]: The above README comments were out of date.

File: lib/lev/background_job.rb [hunk 2]
```diff
@@ -23,7 +23,7 @@ class BackgroundJob
     def initialize(attrs = {})
       @id = attrs[:id] || attrs['id'] || SecureRandom.uuid
       @status = attrs[:status] || attrs['status'] || STATE_UNKNOWN
-      @progress = attrs[:progress] || attrs['progress'] || set_progress(0)
+      @progress = attrs[:progress] || attrs['progress'] || 0
```
Comment (jpslav) [inline lib/lev/background_job.rb]: `set_progress` was just one more call to `set` which was not needed.

File: lib/lev/background_job.rb [hunk 3]
```diff
@@ -35,8 +35,8 @@ def initialize(attrs = {})
     def self.find(id)
       attrs = { id: id }
 
-      if job = store.fetch(job_key(id))
-        attrs.merge!(JSON.parse(job))
+      if job = fetch_and_parse(job_key(id))
```
Comment (jpslav) [inline lib/lev/background_job.rb]: `fetch_and_parse` makes it so we only parse (and stringify keys) in one place

File: lib/lev/background_job.rb [hunk 4]
```diff
@@ -113,22 +119,25 @@ def self.store
...
-      end
+      self.class.fetch_and_parse(job_key) || {}
     end
 
     def track_job_id
       ids = self.class.job_ids
+      return if ids.include?(@id)
       ids << @id
-      self.class.store.write(self.class.job_key('lev_job_ids'), ids.uniq)
+      self.class.store.write(self.class.job_key('lev_job_ids'), ids)
```
Comment (jpslav) [inline lib/lev/background_job.rb]: The before and after here may look equivalent, but the prior approach gave us different string objects from call to call even if the ID already existed.  This caused the `all` implementation to return duplicate `BackgroundJob`s.

File: spec/lev/status_spec.rb [hunk 5]
```diff
@@ -1,44 +0,0 @@
-require 'spec_helper'
```
Comment (jpslav) [inline spec/lev/status_spec.rb]: This spec moved into the background job spec

**Comment by jpslav** [comment]
This may have to be a major version change since I changed the exception type when raising fatal errors

**Comment by joemsak** [comment]
:+1:

**Pull request merged**